onnxruntime/test/optimizer/nchwc_optimizer_test.cc

```diff
@@ -0,0 +1,839 @@
+      auto* output_arg = helper.MakeOutput();
+
+      auto* conv_output_arg = output_arg;
+      if (!activation_op_type.empty()) {
+        conv_output_arg = helper.MakeIntermediate();
+        helper.AddNode(activation_op_type, {conv_output_arg}, {output_arg});
+      }
+
+      auto& conv_node = helper.AddConvNode(input_arg, conv_output_arg, {96, 1, 3, 3});
+      conv_node.AddAttribute("group", static_cast<int64_t>(96));
+    };
+
+    auto check_nchwc_graph = [&](NchwcInferenceSession& session) {
+      auto op_to_count = session.CountOpsInGraph();
+      EXPECT_EQ(op_to_count["nchwc.Conv"], 1);
+      EXPECT_EQ(op_to_count["nchwc.ReorderInput"], 1);
+      EXPECT_EQ(op_to_count["nchwc.ReorderOutput"], 1);
+      if (!activation_op_type.empty()) {
+        EXPECT_EQ(op_to_count[activation_op_type], 0);
+      }
+    };
+
+    NchwcOptimizerTester(build_test_case, check_nchwc_graph);
+  };
+
+  std::vector<std::string> activation_op_types = {"", "Relu", "LeakyRelu"};
+  for (auto& activation_op_type : activation_op_types) {
+    test_case(activation_op_type);
+  }
+}
+
+TEST(NchwcOptimizerTests, ConvPointwise) {
+  auto test_case = [&](const std::string& activation_op_type) {
+    auto build_test_case = [&](NchwcTestHelper& helper) {
+      auto* input_arg = helper.MakeInput({16, 64, 28, 42});
+      auto* output_arg = helper.MakeOutput();
+
+      auto* conv_output_arg = output_arg;
+      if (!activation_op_type.empty()) {
+        conv_output_arg = helper.MakeIntermediate();
+        helper.AddNode(activation_op_type, {conv_output_arg}, {output_arg});
+      }
+
+      helper.AddConvNode(input_arg, conv_output_arg, {128, 64, 1, 1});
+    };
+
+    auto check_nchwc_graph = [&](NchwcInferenceSession& session) {
+      auto op_to_count = session.CountOpsInGraph();
+      EXPECT_EQ(op_to_count["nchwc.Conv"], 1);
+      EXPECT_EQ(op_to_count["nchwc.ReorderInput"], 1);
+      EXPECT_EQ(op_to_count["nchwc.ReorderOutput"], 1);
+      if (!activation_op_type.empty()) {
+        EXPECT_EQ(op_to_count[activation_op_type], 0);
+      }
+    };
+
+    NchwcOptimizerTester(build_test_case, check_nchwc_graph);
+  };
+
+  std::vector<std::string> activation_op_types = {"", "Relu", "LeakyRelu"};
+  for (auto& activation_op_type : activation_op_types) {
+    test_case(activation_op_type);
+  }
+}
+
+TEST(NchwcOptimizerTests, ConvClip) {
+  auto build_test_case = [&](NchwcTestHelper& helper) {
+    auto* input_arg = helper.MakeInput({1, 3, 28, 28});
+    auto* conv_output_arg = helper.MakeIntermediate();
+    auto* output_arg = helper.MakeOutput();
+
+    helper.AddConvNode(input_arg, conv_output_arg, {128, 3, 1, 1});
+
+    auto& clip_node = helper.AddNode("Clip", {conv_output_arg}, {output_arg});
+    clip_node.AddAttribute("min", 0.0f);
+    clip_node.AddAttribute("max", 6.0f);
+  };
+
+  auto check_nchwc_graph = [&](NchwcInferenceSession& session) {
+    auto op_to_count = session.CountOpsInGraph();
+    EXPECT_EQ(op_to_count["nchwc.Conv"], 1);
```

Comment:
There's no fusion of the Clip yet is there? Just wondering as there's no test that there's still a Clip node.

Reply:
The point of the test was to exercise the Clip path in the transformer which keeps the inputs/outputs in NCHWc format to avoid unnecessary reordering, but the test also didn't do a good job of fulling exercising this. Now, the Clip is between two Conv nodes and the whole sequence shouldn't have any ReorderInputs present.